etna/pipeline/base.py

```diff
@@ -24,6 +25,87 @@
 from etna.metrics import Metric
 from etna.metrics import MetricAggregationMode
 
+Timestamp = Union[str, pd.Timestamp]
+
+
+class CrossValidationMode(Enum):
+    """Enum for different cross-validation modes."""
+
+    expand = "expand"
+    constant = "constant"
+
+
+class FoldMask(BaseMixin):
+    """FoldMask class for masks."""
+
+    def __init__(
+        self,
+        first_train_timestamp: Optional[Timestamp],
+        last_train_timestamp: Timestamp,
+        target_timestamps: List[Timestamp],
+    ):
+        """Init FoldMask.
+
+        Parameters
+        ----------
+        first_train_timestamp:
+            First train timestamp, the first timestamp in the dataset if None is passed
+        last_train_timestamp:
+            Last train timestamp
+        target_timestamps:
+            List of target timestamps
+        """
+        self.first_train_timestamp = pd.to_datetime(first_train_timestamp) if first_train_timestamp else None
+        self.last_train_timestamp = pd.to_datetime(last_train_timestamp)
+        self.target_timestamps = sorted([pd.to_datetime(timestamp) for timestamp in target_timestamps])
+
+        self._validate_last_train_timestamp()
+        self._validate_target_timestamps()
+
+    def _validate_last_train_timestamp(self):
+        """Check that last train timestamp is later then first train timestamp."""
+        if self.first_train_timestamp and self.last_train_timestamp < self.first_train_timestamp:
+            raise ValueError("Last train timestamp should be strictly later than first train timestamp!")
+
+    def _validate_target_timestamps(self):
+        """Check that all target timestamps are later then last train timestamp."""
+        first_target_timestamp = self.target_timestamps[0]
+        if first_target_timestamp < self.last_train_timestamp:
+            raise ValueError("Target timestamps should be strictly later then last train timestamp!")
+
+    def validate_on_dataset(self, ts: TSDataset, horizon: int):
+        """Validate fold mask on the dataset with specified horizon.
+
+        Parameters
+        ----------
+        ts:
+            Dataset to validate on
+        horizon:
+            Forecasting horizon
+        """
+        dataset_timestamps = list(ts.index)
+        dataset_description = ts.describe()
+
+        min_first_timestamp = dataset_description["start_timestamp"].min()
+        if self.first_train_timestamp and self.first_train_timestamp < min_first_timestamp:
+            raise ValueError(f"First train timestamp should be later than {min_first_timestamp}!")
+
+        last_timestamp = dataset_description["end_timestamp"].min()
+        if self.last_train_timestamp > last_timestamp:
+            raise ValueError(f"Last train timestamp should be sooner than {last_timestamp}!")
```

Comment:
Not precise error message. The error occurs only if last_train_timestamp is later that last_timestamp. It means that last_train_timestamp should be not later then last_timestamp, not sooner.